File: web/src/api/agent-task.ts
/**
 * Agent Task API
 *
 * Provides API functions for managing scheduled agent tasks, including
 * CRUD operations, status management, and event retrieval.
 */

import { apiClient, isTauri, tauriInvoke } from './config'
import { listen, type UnlistenFn } from '@tauri-apps/api/event'
import type { AgentTask } from '@/types/generated/AgentTask'
import type { AgentTaskStatus } from '@/types/generated/AgentTaskStatus'
import type { TaskEvent } from '@/types/generated/TaskEvent'
import type { TaskSchedule } from '@/types/generated/TaskSchedule'
import type { NotificationConfig } from '@/types/generated/NotificationConfig'
import type { ExecutionMode } from '@/types/generated/ExecutionMode'
import type { MemoryConfig } from '@/types/generated/MemoryConfig'
import type { MemoryScope } from '@/types/generated/MemoryScope'
import type { TaskStreamEvent } from '@/types/generated/TaskStreamEvent'
import type { StreamEventKind } from '@/types/generated/StreamEventKind'
import type { BackgroundAgentControlAction } from '@/types/generated/BackgroundAgentControlAction'
import type { BackgroundProgress } from '@/types/generated/BackgroundProgress'
import type { BackgroundMessage } from '@/types/generated/BackgroundMessage'
import type { BackgroundMessageSource } from '@/types/generated/BackgroundMessageSource'
import { API_ENDPOINTS } from '@/constants'

export interface ActiveTaskInfo {
  task_id: string
  task_name: string
  agent_id: string
  started_at: number
  execution_mode: string
}

// Re-export types for convenience
export type {
  AgentTask,
  AgentTaskStatus,
  TaskEvent,
  TaskSchedule,
  NotificationConfig,
  ExecutionMode,
  MemoryConfig,
  MemoryScope,
  TaskStreamEvent,
  StreamEventKind,
  BackgroundAgentControlAction,
  BackgroundProgress,
  BackgroundMessage,
  BackgroundMessageSource,
}

/**
 * Event name for task stream events (matches Rust constant TASK_STREAM_EVENT)
 */
export const TASK_STREAM_EVENT = 'background-agent:stream'

/**
 * Request to create a new agent task
 */
export interface CreateAgentTaskRequest {
  /** Display name of the task */
  name: string
  /** ID of the agent to execute */
  agent_id: string
  /** Schedule configuration */
  schedule: TaskSchedule
  /** Optional description */
  description?: string
  /** Optional input/prompt to send to the agent */
  input?: string
  /** Optional runtime template used to render task input */
  input_template?: string
  /** Optional notification configuration */
  notification?: NotificationConfig
  /** Optional execution mode (API or CLI) */
  execution_mode?: ExecutionMode
  /** Optional memory configuration */
  memory?: MemoryConfig
  /** Optional memory scope override */
  memory_scope?: MemoryScope
}

/**
 * Request to update an existing agent task
 */
export interface UpdateAgentTaskRequest {
  /** New display name (optional) */
  name?: string
  /** New description (optional) */
  description?: string
  /** New agent ID (optional) */
  agent_id?: string
  /** New input/prompt (optional) */
  input?: string
  /** New runtime template (optional) */
  input_template?: string
  /** New schedule (optional) */
  schedule?: TaskSchedule
  /** New notification config (optional) */
  notification?: NotificationConfig
  /** New memory config (optional) */
  memory?: MemoryConfig
  /** New memory scope override (optional) */
  memory_scope?: MemoryScope
}

interface ControlBackgroundAgentRequest {
  action: BackgroundAgentControlAction
}

interface SendBackgroundMessageRequest {
  message: string
  source?: BackgroundMessageSource
}

/**
 * List all agent tasks
 */
export async function listAgentTasks(): Promise<AgentTask[]> {
  if (isTauri()) {
    return tauriInvoke<AgentTask[]>('list_agent_tasks')
  }
  const response = await apiClient.get<AgentTask[]>(API_ENDPOINTS.AGENT_TASK.LIST)
  return response.data
}

/**
 * List agent tasks filtered by status
 */
export async function listAgentTasksByStatus(status: AgentTaskStatus): Promise<AgentTask[]> {
  if (isTauri()) {
    return tauriInvoke<AgentTask[]>('list_agent_tasks_by_status', { status })
  }
  const response = await apiClient.get<AgentTask[]>(API_ENDPOINTS.AGENT_TASK.LIST_BY_STATUS(status))
  return response.data
}

/**
 * Get an agent task by ID
 */
export async function getAgentTask(id: string): Promise<AgentTask> {
  if (isTauri()) {
    return tauriInvoke<AgentTask>('get_agent_task', { id })
  }
  const response = await apiClient.get<AgentTask>(API_ENDPOINTS.AGENT_TASK.GET(id))
  return response.data
}

/**
 * Create a new agent task
 */
export async function createAgentTask(request: CreateAgentTaskRequest): Promise<AgentTask> {
  if (isTauri()) {
    return tauriInvoke<AgentTask>('create_agent_task', { request })
  }
  const response = await apiClient.post<AgentTask>(API_ENDPOINTS.AGENT_TASK.CREATE, request)
  return response.data
}

/**
 * Update an existing agent task
 */
export async function updateAgentTask(id: string, request: UpdateAgentTaskRequest): Promise<AgentTask> {
  if (isTauri()) {
    return tauriInvoke<AgentTask>('update_agent_task', { id, request })
  }
  const response = await apiClient.patch<AgentTask>(API_ENDPOINTS.AGENT_TASK.UPDATE(id), request)
  return response.data
}

/**
 * Delete an agent task
 */
export async function deleteAgentTask(id: string): Promise<boolean> {
  if (isTauri()) {
    return tauriInvoke<boolean>('delete_agent_task', { id })
  }
  const response = await apiClient.delete<boolean>(API_ENDPOINTS.AGENT_TASK.DELETE(id))
  return response.data
}

/**
 * Pause an agent task
 */
export async function pauseAgentTask(id: string): Promise<AgentTask> {
  if (isTauri()) {
    return tauriInvoke<AgentTask>('pause_agent_task', { id })
  }
  const body: ControlBackgroundAgentRequest = { action: 'pause' }
  const response = await apiClient.post<AgentTask>(API_ENDPOINTS.AGENT_TASK.CONTROL(id), body)
  return response.data
}

/**
 * Resume a paused agent task
 */
export async function resumeAgentTask(id: string): Promise<AgentTask> {
  if (isTauri()) {
    return tauriInvoke<AgentTask>('resume_agent_task', { id })
  }
  const body: ControlBackgroundAgentRequest = { action: 'resume' }
  const response = await apiClient.post<AgentTask>(API_ENDPOINTS.AGENT_TASK.CONTROL(id), body)
  return response.data
}

/**
 * Get events for a specific task
 * @param taskId - The task ID
 * @param limit - Optional maximum number of events to return
 */
export async function getAgentTaskEvents(taskId: string, limit?: number): Promise<TaskEvent[]> {
  if (isTauri()) {
    return tauriInvoke<TaskEvent[]>('get_agent_task_events', { taskId, limit })
  }
  const url = limit
    ? `${API_ENDPOINTS.AGENT_TASK.PROGRESS(taskId)}?event_limit=${limit}`
    : API_ENDPOINTS.AGENT_TASK.PROGRESS(taskId)
  const response = await apiClient.get<BackgroundProgress>(url)
  return response.data.recent_events
}

/**
 * Get tasks that are ready to run (based on their schedule)
 */
export async function getRunnableAgentTasks(): Promise<AgentTask[]> {
  if (isTauri()) {
    return tauriInvoke<AgentTask[]>('get_runnable_agent_tasks')
  }
  const response = await apiClient.get<AgentTask[]>(API_ENDPOINTS.AGENT_TASK.LIST_BY_STATUS('active'))
  const now = Date.now()
  return response.data.filter(
    task =>
      task.status === 'active'
      && (task.next_run_at === null || task.next_run_at <= now),
  )
}

/**
 * Send an interaction message to a background agent.
 */
export async function sendBackgroundAgentMessage(
  id: string,
  request: SendBackgroundMessageRequest,
): Promise<BackgroundMessage> {
  const response = await apiClient.post<BackgroundMessage>(API_ENDPOINTS.AGENT_TASK.MESSAGES(id), request)
  return response.data
}

/**
 * List recent interaction messages for a background agent.
 */
export async function listBackgroundAgentMessages(
  id: string,
  limit = 50,
): Promise<BackgroundMessage[]> {
  const response = await apiClient.get<BackgroundMessage[]>(
    `${API_ENDPOINTS.AGENT_TASK.MESSAGES(id)}?limit=${limit}`,
  )
  return response.data
}

/**
 * Helper to create a default notification config
 */
export function createDefaultNotificationConfig(): NotificationConfig {
  return {
    telegram_enabled: false,
    telegram_bot_token: null,
    telegram_chat_id: null,
    notify_on_failure_only: false,
    include_output: true,
  }
}

/**
 * Helper to create a one-time schedule
 * @param runAt - Unix timestamp in milliseconds
 */
export function createOnceSchedule(runAt: number): TaskSchedule {
  return {
    type: 'once',
    run_at: runAt,
  }
}

/**
 * Helper to create an interval schedule
 * @param intervalMs - Interval in milliseconds
 * @param startAt - Optional start time (defaults to now)
 */
export function createIntervalSchedule(intervalMs: number, startAt?: number): TaskSchedule {
  return {
    type: 'interval',
    interval_ms: intervalMs,
    start_at: startAt ?? null,
  }
}

/**
 * Helper to create a cron schedule
 * @param expression - Cron expression (e.g., "0 9 * * *")
 * @param timezone - Optional timezone (e.g., "America/Los_Angeles")
 */
export function createCronSchedule(expression: string, timezone?: string): TaskSchedule {
  return {
    type: 'cron',
    expression,
    timezone: timezone ?? null,
  }
}

/**
 * Format schedule for display
 */
export function formatSchedule(schedule: TaskSchedule): string {
  switch (schedule.type) {
    case 'once':
      return `Once at ${new Date(schedule.run_at).toLocaleString()}`
    case 'interval': {
      const hours = Math.floor(schedule.interval_ms / 3600000)
      const minutes = Math.floor((schedule.interval_ms % 3600000) / 60000)
      if (hours > 0 && minutes > 0) {
        return `Every ${hours}h ${minutes}m`
      } else if (hours > 0) {
        return `Every ${hours} hour${hours > 1 ? 's' : ''}`
      } else {
        return `Every ${minutes} minute${minutes > 1 ? 's' : ''}`
      }
    }
    case 'cron':
      return `Cron: ${schedule.expression}${schedule.timezone ? ` (${schedule.timezone})` : ''}`
    default:
      return 'Unknown schedule'
  }
}

/**
 * Format task status for display
 */
export function formatTaskStatus(status: AgentTaskStatus): string {
  const statusMap: Record<AgentTaskStatus, string> = {
    active: 'Active',
    paused: 'Paused',
    running: 'Running',
    completed: 'Completed',
    failed: 'Failed',
  }
  return statusMap[status] || status
}

/**
 * Get status badge color class
 */
export function getStatusColor(status: AgentTaskStatus): string {
  const colorMap: Record<AgentTaskStatus, string> = {
    active: 'success',
    paused: 'info',
    running: 'primary',
    completed: 'success',
    failed: 'danger',
  }
  return colorMap[status] || 'default'
}

// ============================================================================
// Task Stream Event Handling
// ============================================================================

/**
 * Listen for all task stream events
 *
 * This registers a listener for all task execution events. Use this for
 * global monitoring or when managing multiple concurrent tasks.
 *
 * @param callback - Function to call with each event
 * @returns Unlisten function to stop listening
 */
export async function onTaskStreamEvent(
  callback: (event: TaskStreamEvent) => void,
): Promise<UnlistenFn> {
  if (!isTauri()) {
    // In web mode, return a no-op unlisten function
    console.warn('Task stream events are only available in Tauri desktop app')
    return () => {}
  }

  return listen<TaskStreamEvent>(TASK_STREAM_EVENT, (event) => {
    callback(event.payload)
  })
}

/**
 * Listen for stream events for a specific task
 *
 * This filters events to only those belonging to the specified task.
 * Use this when monitoring a single task's execution.
 *
 * @param taskId - Task ID to filter events
 * @param callback - Function to call with each matching event
 * @returns Unlisten function to stop listening
 */
export async function onTaskStreamEventForTask(
  taskId: string,
  callback: (event: TaskStreamEvent) => void,
): Promise<UnlistenFn> {
  if (!isTauri()) {
    console.warn('Task stream events are only available in Tauri desktop app')
    return () => {}
  }

  return listen<TaskStreamEvent>(TASK_STREAM_EVENT, (event) => {
    if (event.payload.task_id === taskId) {
      callback(event.payload)
    }
  })
}

/**
 * Type guard for checking event kind
 */
export function isEventKind<T extends StreamEventKind['type']>(
  event: TaskStreamEvent,
  type: T,
): event is TaskStreamEvent & { kind: Extract<StreamEventKind, { type: T }> } {
  return event.kind.type === type
}

/**
 * Run a task with streaming support
 *
 * This is a convenience wrapper that invokes the Tauri command to run
 * a task with real-time event streaming.
 *
 * @param taskId - ID of the task to run
 * @returns Promise that resolves when the task starts (not when it completes)
 */
export async function runAgentTaskStreaming(taskId: string): Promise<void> {
  if (!isTauri()) {
    throw new Error('Streaming task execution is only available in Tauri desktop app')
  }
  return tauriInvoke<void>('run_agent_task_streaming', { id: taskId })
}

/**
 * Get list of currently active (running) task IDs
 *
 * @returns Array of task IDs that are currently running
 */
export async function getActiveAgentTasks(): Promise<ActiveTaskInfo[]> {
  if (!isTauri()) {
    return []
  }
  return tauriInvoke<ActiveTaskInfo[]>('get_active_agent_tasks')
}

/**
 * Cancel a running task
 *
 * @param taskId - ID of the task to cancel
 * @returns true if cancellation was requested successfully
 */
export async function cancelAgentTask(taskId: string): Promise<boolean> {
  if (!isTauri()) {
    throw new Error('Task cancellation is only available in Tauri desktop app')
  }
  return tauriInvoke<boolean>('cancel_agent_task', { taskId })
}

File: crates/restflow-tauri/src/commands/agent_task.rs
//! Agent Task Tauri commands
//!
//! Provides IPC commands for managing scheduled agent tasks from the frontend.
//!
//! # Streaming Events
//!
//! The `run_background_agent_streaming` command executes a task immediately and streams
//! real-time events to the frontend via Tauri's event system. Frontend should
//! listen to the `background-agent:stream` event to receive `TaskStreamEvent` updates.
//!
//! ```typescript
//! import { listen } from '@tauri-apps/api/event';
//! import type { TaskStreamEvent } from './types/generated';
//!
//! const unlisten = await listen<TaskStreamEvent>('background-agent:stream', (event) => {
//!   console.log('Task event:', event.payload);
//! });
//! ```

use crate::agent_task::{TASK_STREAM_EVENT, TaskStreamEvent, TauriEventEmitter};
use crate::state::AppState;
use restflow_core::models::{
    AgentTask, AgentTaskStatus, BackgroundAgentControlAction, BackgroundAgentPatch,
    BackgroundAgentSpec, ExecutionMode, MemoryConfig, MemoryScope, NotificationConfig,
    SteerMessage, SteerSource, TaskEvent, TaskSchedule,
};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, State};

/// Request to create a new agent task
#[derive(Debug, Deserialize)]
pub struct CreateBackgroundAgentRequest {
    /// Display name of the task
    pub name: String,
    /// ID of the agent to execute
    pub agent_id: String,
    /// Schedule configuration
    pub schedule: TaskSchedule,
    /// Optional description
    #[serde(default)]
    pub description: Option<String>,
    /// Optional input/prompt to send to the agent
    #[serde(default)]
    pub input: Option<String>,
    /// Optional runtime template for constructing input
    #[serde(default)]
    pub input_template: Option<String>,
    /// Optional notification configuration
    #[serde(default)]
    pub notification: Option<NotificationConfig>,
    /// Optional execution mode (API or CLI)
    #[serde(default)]
    pub execution_mode: Option<ExecutionMode>,
    /// Optional memory configuration
    #[serde(default)]
    pub memory: Option<MemoryConfig>,
    /// Optional memory scope override
    #[serde(default)]
    pub memory_scope: Option<MemoryScope>,
}

/// Request to update an existing agent task
#[derive(Debug, Deserialize)]
pub struct UpdateBackgroundAgentRequest {
    /// New display name (optional)
    #[serde(default)]
    pub name: Option<String>,
    /// New description (optional)
    #[serde(default)]
    pub description: Option<String>,
    /// New agent ID (optional)
    #[serde(default)]
    pub agent_id: Option<String>,
    /// New input/prompt (optional)
    #[serde(default)]
    pub input: Option<String>,
    /// New runtime input template (optional)
    #[serde(default)]
    pub input_template: Option<String>,
    /// New schedule (optional)
    #[serde(default)]
    pub schedule: Option<TaskSchedule>,
    /// New notification config (optional)
    #[serde(default)]
    pub notification: Option<NotificationConfig>,
    /// New memory configuration (optional)
    #[serde(default)]
    pub memory: Option<MemoryConfig>,
    /// New memory scope override (optional)
    #[serde(default)]
    pub memory_scope: Option<MemoryScope>,
}

/// List all agent tasks
#[tauri::command]
pub async fn list_background_agents(state: State<'_, AppState>) -> Result<Vec<AgentTask>, String> {
    state
        .executor()
        .list_background_agents(None)
        .await
        .map_err(|e| e.to_string())
}

/// List agent tasks filtered by status
#[tauri::command]
pub async fn list_background_agents_by_status(
    state: State<'_, AppState>,
    status: AgentTaskStatus,
) -> Result<Vec<AgentTask>, String> {
    let status_str = match status {
        AgentTaskStatus::Active => "active",
        AgentTaskStatus::Paused => "paused",
        AgentTaskStatus::Running => "running",
        AgentTaskStatus::Completed => "completed",
        AgentTaskStatus::Failed => "failed",
    };
    state
        .executor()
        .list_background_agents(Some(status_str.to_string()))
        .await
        .map_err(|e| e.to_string())
}

/// Get an agent task by ID
#[tauri::command]
pub async fn get_background_agent(
    state: State<'_, AppState>,
    id: String,
) -> Result<AgentTask, String> {
    state
        .executor()
        .get_background_agent(id.clone())
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Background agent '{}' not found", id))
}

/// Create a new agent task
#[tauri::command]
pub async fn create_background_agent(
    state: State<'_, AppState>,
    request: CreateBackgroundAgentRequest,
) -> Result<AgentTask, String> {
    let spec = BackgroundAgentSpec {
        name: request.name,
        agent_id: request.agent_id,
        description: request.description,
        input: request.input,
        input_template: request.input_template,
        schedule: request.schedule,
        notification: request.notification,
        execution_mode: request.execution_mode,
        memory: merge_memory_scope(request.memory, request.memory_scope),
    };

    state
        .executor()
        .create_background_agent(spec)
        .await
        .map_err(|e| e.to_string())
}

/// Update an existing agent task
#[tauri::command]
pub async fn update_background_agent(
    state: State<'_, AppState>,
    id: String,
    request: UpdateBackgroundAgentRequest,
) -> Result<AgentTask, String> {
    let patch = BackgroundAgentPatch {
        name: request.name,
        description: request.description,
        agent_id: request.agent_id,
        input: request.input,
        input_template: request.input_template,
        schedule: request.schedule,
        notification: request.notification,
        execution_mode: None,
        memory: merge_memory_scope(request.memory, request.memory_scope),
    };

    state
        .executor()
        .update_background_agent(id, patch)
        .await
        .map_err(|e| e.to_string())
}

/// Delete an agent task
#[tauri::command]
pub async fn delete_background_agent(
    state: State<'_, AppState>,
    id: String,
) -> Result<bool, String> {
    state
        .executor()
        .delete_background_agent(id)
        .await
        .map_err(|e| e.to_string())
}

/// Pause an agent task
#[tauri::command]
pub async fn pause_background_agent(
    state: State<'_, AppState>,
    id: String,
) -> Result<AgentTask, String> {
    state
        .executor()
        .control_background_agent(id, BackgroundAgentControlAction::Pause)
        .await
        .map_err(|e| e.to_string())
}

/// Resume a paused agent task
#[tauri::command]
pub async fn resume_background_agent(
    state: State<'_, AppState>,
    id: String,
) -> Result<AgentTask, String> {
    state
        .executor()
        .control_background_agent(id, BackgroundAgentControlAction::Resume)
        .await
        .map_err(|e| e.to_string())
}

/// Cancel a running agent task
#[tauri::command]
pub async fn cancel_background_agent(
    state: State<'_, AppState>,
    task_id: String,
) -> Result<bool, String> {
    state
        .cancel_task(task_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(true)
}

/// Send a steer message to a running task.
///
/// This allows injecting new instructions into a running agent's ReAct loop.
/// The instruction will be processed at the next iteration of the loop.
///
/// Returns true if the message was sent, false if the task is not running
/// or doesn't support steering (e.g., CLI execution mode).
#[tauri::command]
pub async fn steer_task(
    state: State<'_, AppState>,
    task_id: String,
    instruction: String,
) -> Result<bool, String> {
    let message = SteerMessage {
        instruction,
        source: SteerSource::User,
        timestamp: chrono::Utc::now().timestamp_millis(),
    };

    let sent = state.steer_registry.steer(&task_id, message).await;
    Ok(sent)
}

/// Get events for a specific task
#[tauri::command]
pub async fn get_background_agent_events(
    state: State<'_, AppState>,
    task_id: String,
    limit: Option<usize>,
) -> Result<Vec<TaskEvent>, String> {
    let mut events = state
        .executor()
        .get_background_agent_history(task_id)
        .await
        .map_err(|e| e.to_string())?;

    // Apply limit if specified (IPC doesn't support limit natively)
    if let Some(limit) = limit {
        events.truncate(limit);
    }

    Ok(events)
}

/// Get runnable tasks (tasks that should run now based on schedule)
#[tauri::command]
pub async fn get_runnable_background_agents(
    state: State<'_, AppState>,
) -> Result<Vec<AgentTask>, String> {
    let current_time = chrono::Utc::now().timestamp_millis();
    let core = state.core.as_ref().ok_or("AppCore not available")?;
    core.storage
        .agent_tasks
        .list_runnable_tasks(current_time)
        .map_err(|e| e.to_string())
}

fn merge_memory_scope(
    memory: Option<MemoryConfig>,
    memory_scope: Option<MemoryScope>,
) -> Option<MemoryConfig> {
    match (memory, memory_scope) {
        (Some(mut memory), Some(scope)) => {
            memory.memory_scope = scope;
            Some(memory)
        }
        (Some(memory), None) => Some(memory),
        (None, Some(scope)) => Some(MemoryConfig {
            memory_scope: scope,
            ..MemoryConfig::default()
        }),
        (None, None) => None,
    }
}

// ============================================================================
// Streaming Event Commands
// ============================================================================

/// Response for streaming task execution
#[derive(Debug, Clone, Serialize)]
pub struct StreamingBackgroundAgentResponse {
    /// Task ID that was started
    pub task_id: String,
    /// Event channel name to listen on
    pub event_channel: String,
    /// Whether the task is already running (queued)
    pub already_running: bool,
}

/// Run an agent task immediately and stream events to the frontend.
///
/// This command triggers immediate execution of a task and emits real-time
/// events via Tauri's event system. The frontend should listen to the
/// `background-agent:stream` event to receive `TaskStreamEvent` updates.
///
/// # Arguments
///
/// * `id` - The ID of the agent task to run
///
/// # Returns
///
/// Returns a `StreamingBackgroundAgentResponse` with the task ID and event channel name.
///
/// # Events
///
/// The following events are emitted on the `background-agent:stream` channel:
/// - `started` - Task execution has begun
/// - `output` - Output from the task (stdout/stderr)
/// - `progress` - Progress updates for long-running tasks
/// - `completed` - Task finished successfully
/// - `failed` - Task failed with an error
/// - `cancelled` - Task was cancelled (timeout or user request)
/// - `heartbeat` - Periodic heartbeat while task is running
///
/// # Example (Frontend)
///
/// ```typescript
/// import { invoke } from '@tauri-apps/api/core';
/// import { listen } from '@tauri-apps/api/event';
///
/// // Start listening before invoking
/// const unlisten = await listen<TaskStreamEvent>('background-agent:stream', (event) => {
///   if (event.payload.task_id === taskId) {
///     switch (event.payload.kind.type) {
///       case 'started':
///         console.log('Task started:', event.payload.kind.task_name);
///         break;
///       case 'output':
///         console.log('Output:', event.payload.kind.text);
///         break;
///       case 'completed':
///         console.log('Done:', event.payload.kind.result);
///         break;
///     }
///   }
/// });
///
/// // Trigger task execution
/// const response = await invoke('run_background_agent_streaming', { id: 'task-123' });
/// console.log('Started task:', response.task_id);
/// ```
#[tauri::command]
pub async fn run_background_agent_streaming(
    state: State<'_, AppState>,
    app_handle: AppHandle,
    id: String,
) -> Result<StreamingBackgroundAgentResponse, String> {
    // Check if task exists
    let core = state.core.as_ref().ok_or("AppCore not available")?;
    let task = core
        .storage
        .agent_tasks
        .get_task(&id)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Background agent '{}' not found", id))?;

    // Check if already running
    let already_running = state.is_task_running(&id).await;
    if already_running {
        return Ok(StreamingBackgroundAgentResponse {
            task_id: id,
            event_channel: TASK_STREAM_EVENT.to_string(),
            already_running: true,
        });
    }

    // Emit started event (emitter stored for future use in enhanced streaming)
    let _emitter = TauriEventEmitter::new(app_handle.clone());
    let execution_mode_str = match &task.execution_mode {
        ExecutionMode::Api => "api".to_string(),
        ExecutionMode::Cli(cfg) => format!("cli:{}", cfg.binary),
    };

    let started_event =
        TaskStreamEvent::started(&task.id, &task.name, &task.agent_id, &execution_mode_str);

    if let Err(e) = app_handle.emit(TASK_STREAM_EVENT, &started_event) {
        tracing::warn!("Failed to emit started event: {}", e);
    }

    // Trigger the task execution via runner (which will emit more events)
    if let Err(e) = state.run_task_now(id.clone()).await {
        // Emit failed event
        let failed_event = TaskStreamEvent::failed(&id, e.to_string(), 0, false);
        let _ = app_handle.emit(TASK_STREAM_EVENT, &failed_event);
        return Err(e.to_string());
    }

    Ok(StreamingBackgroundAgentResponse {
        task_id: id,
        event_channel: TASK_STREAM_EVENT.to_string(),
        already_running: false,
    })
}

/// Information about an active/running task
#[derive(Debug, Clone, Serialize)]
pub struct ActiveBackgroundAgentInfo {
    /// Task ID
    pub task_id: String,
    /// Task name
    pub task_name: String,
    /// Agent ID being executed
    pub agent_id: String,
    /// When the task started (milliseconds since epoch)
    pub started_at: i64,
    /// Execution mode
    pub execution_mode: String,
}

/// Get list of currently running/active agent tasks
///
/// Returns information about all tasks that are currently being executed.
#[tauri::command]
pub async fn get_active_background_agents(
    state: State<'_, AppState>,
) -> Result<Vec<ActiveBackgroundAgentInfo>, String> {
    state.get_active_tasks().await.map_err(|e| e.to_string())
}

/// Emit a test event to verify the streaming system is working
///
/// This is useful for debugging and testing the event system from the frontend.
#[tauri::command]
pub async fn emit_test_background_agent_event(
    app_handle: AppHandle,
    task_id: String,
    message: String,
) -> Result<(), String> {
    let event = TaskStreamEvent::output(&task_id, &message, false);
    app_handle
        .emit(TASK_STREAM_EVENT, &event)
        .map_err(|e| e.to_string())
}

/// Subscribe to task stream events
///
/// This is a no-op command that documents the event subscription pattern.
/// In Tauri v2, the frontend uses `listen()` to subscribe to events.
///
/// # Usage
///
/// ```typescript
/// import { listen } from '@tauri-apps/api/event';
/// import type { TaskStreamEvent } from './types/generated';
///
/// // Subscribe to all task events
/// const unlisten = await listen<TaskStreamEvent>('background-agent:stream', (event) => {
///   console.log('Received event:', event.payload);
/// });
///
/// // Later, unsubscribe
/// unlisten();
/// ```
#[tauri::command]
pub fn get_background_agent_stream_event_name() -> String {
    TASK_STREAM_EVENT.to_string()
}

// ============================================================================
// Heartbeat Commands
// ============================================================================

use crate::agent_task::HEARTBEAT_EVENT;

/// Get the heartbeat event name for frontend subscription
///
/// The heartbeat events are now emitted inline by the AgentTaskRunner during
/// its poll cycle, so there's no separate heartbeat runner to manage.
///
/// # Usage
///
/// ```typescript
/// import { listen } from '@tauri-apps/api/event';
/// import type { HeartbeatEvent } from './types/generated';
///
/// const unlisten = await listen<HeartbeatEvent>('background-agent:heartbeat', (event) => {
///   console.log('Heartbeat:', event.payload);
/// });
/// ```
#[tauri::command]
pub fn get_heartbeat_event_name() -> String {
    HEARTBEAT_EVENT.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[allow(dead_code)]
    async fn create_test_state() -> AppState {
        let temp_dir = tempdir().unwrap();
        let db_path = temp_dir.path().join("test.db");
        let db_path_str = db_path.to_str().unwrap();
        AppState::new(db_path_str).await.unwrap()
    }

    #[tokio::test]
    async fn test_create_background_agent_request_deserialize() {
        let json = r#"{
            "name": "Test Task",
            "agent_id": "agent-001",
            "schedule": { "type": "interval", "interval_ms": 3600000 }
        }"#;

        let request: CreateBackgroundAgentRequest = serde_json::from_str(json).unwrap();
        assert_eq!(request.name, "Test Task");
        assert_eq!(request.agent_id, "agent-001");
        assert!(request.description.is_none());
        assert!(request.input.is_none());
        assert!(request.notification.is_none());
    }

    #[tokio::test]
    async fn test_create_background_agent_request_with_all_fields() {
        let json = r#"{
            "name": "Full Task",
            "agent_id": "agent-002",
            "schedule": { "type": "once", "run_at": 1704067200000 },
            "description": "A complete task",
            "input": "Hello agent",
            "notification": {
                "telegram_enabled": true,
                "telegram_chat_id": "123456"
            }
        }"#;

        let request: CreateBackgroundAgentRequest = serde_json::from_str(json).unwrap();
        assert_eq!(request.name, "Full Task");
        assert_eq!(request.agent_id, "agent-002");
        assert_eq!(request.description, Some("A complete task".to_string()));
        assert_eq!(request.input, Some("Hello agent".to_string()));
        assert!(request.notification.is_some());
        let notif = request.notification.unwrap();
        assert!(notif.telegram_enabled);
        assert_eq!(notif.telegram_chat_id, Some("123456".to_string()));
    }

    #[tokio::test]
    async fn test_update_background_agent_request_partial() {
        let json = r#"{
            "name": "New Name"
        }"#;

        let request: UpdateBackgroundAgentRequest = serde_json::from_str(json).unwrap();
        assert_eq!(request.name, Some("New Name".to_string()));
        assert!(request.description.is_none());
        assert!(request.agent_id.is_none());
        assert!(request.input.is_none());
        assert!(request.schedule.is_none());
        assert!(request.notification.is_none());
    }

    #[tokio::test]
    async fn test_cron_schedule_deserialize() {
        let json = r#"{
            "name": "Cron Task",
            "agent_id": "agent-003",
            "schedule": {
                "type": "cron",
                "expression": "0 9 * * *",
                "timezone": "America/Los_Angeles"
            }
        }"#;

        let request: CreateBackgroundAgentRequest = serde_json::from_str(json).unwrap();
        match request.schedule {
            TaskSchedule::Cron {
                expression,
                timezone,
            } => {
                assert_eq!(expression, "0 9 * * *");
                assert_eq!(timezone, Some("America/Los_Angeles".to_string()));
            }
            _ => panic!("Expected Cron schedule"),
        }
    }
}
